_aicspylibczi/Reader.cpp

```diff
@@ -263,18 +267,39 @@ namespace pylibczi {
 
       imageFactory.setMosaic(isMosaic());
       size_t memOffset = 0;
-      for_each(matches.begin(), matches.end(), [&](const SubblockIndexVec::value_type& match_) {
-          auto subblock = m_czireader->ReadSubBlock(match_.second);
-          const libCZI::SubBlockInfo& info = subblock->GetSubBlockInfo();
-          if (m_pixelType!=info.pixelType)
-              throw PixelTypeException(info.pixelType, "Selected subblocks have inconsistent PixelTypes."
-                                                       " You must select subblocks with consistent PixelTypes.");
 
-          imageFactory.constructImage(subblock->CreateBitmap(),
-              &info.coordinate, info.logicalRect, memOffset, info.mIndex);
+      unsigned int number_of_cores = std::thread::hardware_concurrency();
+      if (number_of_cores - 1 < cores_) {
+          std::cout << "Cores exception requested " << cores_ << " but only " << number_of_cores << " available." << std::endl;
+          throw ThreadingRequestedCoresException(number_of_cores, cores_,
+              "Requested cores should be at most 1 less than the number of available cores.");
+      }
+      {
+          std::vector< std::future<bool> > jobs;
+          Tasks tasks;
+          for_each(matches.begin(), matches.end(), [&](const SubblockIndexVec::value_type match_) {
+              int sb_index = match_.second;
+
+              jobs.push_back(tasks.queue([this, &imageFactory, sb_index, memOffset]()->bool {
+                  auto subblock = m_czireader->ReadSubBlock(sb_index);
+                  const libCZI::SubBlockInfo& info = subblock->GetSubBlockInfo();
+                  if (m_pixelType!=info.pixelType)
+                      throw PixelTypeException(info.pixelType, "Selected subblocks have inconsistent PixelTypes."
```

Comment:
what happens if one thread throws but the others don't?  what is the result?

Reply:
Hmm, so the thread throws an exception and then if it's not caught which I don't think it is it would then cause bubble all the way up and cause the program to exit. There's no elegant way to recover. Am I understanding your point correctly or are you saying i need to explicitly handle this exception rather than letting the kernel deal with it?

Reply:
I was wondering if you intend to be handling it at all, if it bubbles up to some catch block near the api call, or what.

Reply:
My thoughts on this are not to handle it. If the user managed to somehow acquire a datafile with varying pixel types then it seems fair to me to let them figure out how to select the consistent data out of the file with the appropriate dims. 

Do you think I should put in a description in the read_image docstring stating that? I have not seen a file with mixed pixel types yet but it is allowed. I honestly don't know how relevant this edge case is in the real world.

Reply:
I think a comment right here where it throws, just to say we are intentional about not catching this upstream, as a conscious choice.  It's just for me, a question of whether it totally crashes the user's app, or whether the user has a shot at catching it and dealing with it themselves.

Reply:
Do throws on the c++ side, that are uncaught in c++, get translated into python exceptions?

Reply:
I have bindings for the C++ exceptions to be cast to python exceptions. It's a nice feature of pybind11
```
py::register_exception<pylibczi::FilePtrException>(m, "PylibCZI_BytesIO2FilePtrException");
py::register_exception<pylibczi::PixelTypeException>(m, "PylibCZI_PixelTypeException");
.
.
.
```